master/pkg/searcher/asha.go

```diff
@@ -0,0 +1,244 @@
+		rungs:              rungs,
+		trialRungs:         make(map[RequestID]int),
+		earlyExitTrials:    make(map[RequestID]bool),
+		maxTrials:          config.MaxTrials,
+	}
+}
+
+// promotions handles bookkeeping of validation metrics and returns a RequestID to promote if
+// appropriate.
+func (r *rung) promotionsAsync(requestID RequestID, metric float64, divisor float64) []RequestID {
+	// See if there is a trial to promote. We are increasing the total number of trials seen by 1; the
+	// number of best trials that definitely should have been promoted so far (numPromote) can only
+	// stay the same or increase by 1.
+	oldNumPromote := int(float64(len(r.metrics)) / divisor)
+	numPromote := int(float64(len(r.metrics)+1) / divisor)
+
+	// Insert the new trial result in the appropriate place in the sorted list.
+	insertIndex := sort.Search(
+		len(r.metrics),
+		func(i int) bool { return r.metrics[i].metric > metric },
+	)
+	promoteNow := insertIndex < numPromote
+
+	r.metrics = append(r.metrics, trialMetric{})
+	copy(r.metrics[insertIndex+1:], r.metrics[insertIndex:])
+	r.metrics[insertIndex] = trialMetric{
+		requestID: requestID,
+		metric:    metric,
+		promoted:  promoteNow,
+		closed:    false,
+	}
+
+	// If the new trial is good enough, it should be promoted immediately (whether or not numPromote
+	// changes). Otherwise, if numPromote changes, there is some other trial that should be promoted,
+	// unless it has been promoted already.
+	if promoteNow {
+		return []RequestID{requestID}
+	}
+	if numPromote != oldNumPromote {
+		t := &r.metrics[oldNumPromote]
+		if !t.promoted {
+			t.promoted = true
+			return []RequestID{t.requestID}
+		}
+	}
+
+	return nil
+}
+
+func (s *asyncHalvingSearch) initialOperations(ctx context) ([]Operation, error) {
+	// The number of initialOperations will control the degree of parallelism
+	// of the search experiment since we guarantee that each validationComplete
+	// call will return a new train workload until we reach MaxTrials.
+
+	// We will use searcher config field if available.
+	// Otherwise we will default to a number of trials that will
+	// guarantee at least one trial at the top rung.
+	var ops []Operation
+	var maxConcurrentTrials int
+
+	if s.MaxConcurrentTrials != nil {
+		maxConcurrentTrials = min(*s.MaxConcurrentTrials, s.MaxTrials)
+	} else {
+		maxConcurrentTrials = max(
+			min(int(math.Pow(s.Divisor, float64(s.NumRungs-1))), s.MaxTrials),
+			1)
+	}
+
+	for trial := 0; trial < maxConcurrentTrials; trial++ {
+		create := NewCreate(
+			ctx.rand, sampleAll(ctx.hparams, ctx.rand), model.TrialWorkloadSequencerType)
+		ops = append(ops, create)
+		ops = append(ops, trainAndValidate(create.RequestID, 0, s.rungs[0].stepsNeeded)...)
+		s.rungs[0].outstandingTrials++
+		s.trialRungs[create.RequestID] = 0
+	}
+	return ops, nil
+}
+
+func (s *asyncHalvingSearch) trainCompleted(
+	ctx context, requestID RequestID, message Workload,
+) ([]Operation, error) {
+	s.totalStepsCompleted++
```

Comment:
question: not sure if this needs a rename or we need to change the logic. Technically we do not know if this is completed since we may want to do a validation and/or checkpoint, and that would happen after we receive this `trainCompleted`

Reply:
trial 1: 0.2
trial 2: 0.3
trial 3: 0.4
trial 4: 0.5
-> promote 1
trial 5: 0.1
-> promote 5
trial 6: 0.6
trial 7: 0.7
trial 8: 0.8
-> nothing to promote even though oldNumPromote < numPromote because trials 1 and 5 already promoted